### PR TITLE
Update the provisioning API docs

### PR DESCRIPTION
**Pull request by pierreca**
Update documentation before the provisioning device SDK release

### CODE REVIEW

File: common/core/src/authentication_provider.ts [hunk 1]
```diff
@@ -3,11 +3,24 @@
 
 import { TransportConfig } from './authorization';
 
+/**
+ * Designate the type of authentication used by an `AuthenticationProvider`.
+ */
 export enum AuthenticationType {
+  /**
+   * X509 authentication uses X509 certificate and key pairs.
```
Comment (anthonyvercolano) [inline common/core/src/authentication_provider.ts]: >uses X509 certificate and key pairs. [](start = 25, length = 36)

Don't really like how this reads.  The x509 certificate itself contains the public key.  The complementary private key is "securely" stored on the device.  The certificate isn't really a certificate without the public key.  Splitting hairs here.

File: common/core/src/authorization.ts [hunk 2]
```diff
@@ -39,21 +39,64 @@ export function hmacHash(password: string, stringToSign: string): string {
   return hmac.digest('base64');
 }
```
Comment (anthonyvercolano) [inline common/core/src/authorization.ts]: The previous 15 lines are a weird color in codeflow.  Don't really see what's causing that.

File: device/core/src/sak_authentication_provider.ts [hunk 3]
```diff
@@ -4,6 +4,13 @@
 import { EventEmitter } from 'events';
 import { AuthenticationProvider, AuthenticationType, ConnectionString, SharedAccessSignature, errors, TransportConfig, encodeUriComponentStrict } from 'azure-iot-common';
 
+/**
+ * Provides an `AuthenticationProvider` object that can be created simply with a connection string and is then used by the device client and transports to authenticate
+ * with the Azure IoT hub instance.
+ *
+ * The `SharedAccessKeyAuthenticationProvider` object takes care of creating shared access signature tokens on a regular cadence and emits the `newTokenAvailable` event for the transports
```
Comment (anthonyvercolano) [inline device/core/src/sak_authentication_provider.ts]: >emits the `newTokenAvailable` event for the transports [](start = 133, length = 54)

I like this.

**Comment by anthonyvercolano** [comment]
>   private _credentials: TransportConfig;

I realize that these are private.  It would still be nice to have documentation about what they are and how they can be used.  Perhaps to keep the documentation private use // comments.

---
Refers to: device/core/src/sak_authentication_provider.ts:23 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

File: provisioning/device/readme.md [hunk 4]
```diff
@@ -1,13 +1,22 @@
-#azure-iot-provisioning-device
-Device client to Communicate with Azure Device Provisioning Service
+This package (**azure-iot-provisioning-device**) provides the device client that is used to communicate with the Azure IoT Hub Device Provisioning Service.
```
Comment (anthonyvercolano) [inline provisioning/device/readme.md]: > Hub  [](start = 122, length = 5)

We are really using the word "hub" prior to the DPS part?

File: provisioning/device/readme.md [hunk 5]
```diff
@@ -1,13 +1,22 @@
...
 
 [![npm version](https://badge.fury.io/js/azure-iot-provisioning-device.svg)](https://badge.fury.io/js/azure-iot-provisioning-device)
 
 ## Install
 
-`npm install -g azure-iot-provisioning-device@latest` to get the latest (pre-release) version.
+There are at least 3 different packages involved when writing code to take advantage of the Azure IoT Hub Device Provisioning Service:
+- The "client" package that will be used to create a provisioning client. This is this package.
```
Comment (anthonyvercolano) [inline provisioning/device/readme.md]: >this  [](start = 82, length = 5)

?This is that?

File: provisioning/device/samples/readme.md [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 This file can be found in https://github.com/Azure/azure-iot-sdk-node/tree/master/provisioning/device/samples
 
-This folder contains simple samples showing how to use the various features of the Microsoft Azure IoT Device Provisioning Service from an application running JavaScript code.
+This folder contains simple samples showing how to use the various features of the Microsoft Azure IoT Device Provisioning Service from an application running JavaScript or TypeScript code.
```
Comment (anthonyvercolano) [inline provisioning/device/samples/readme.md]: >  [](start = 102, length = 1)

Hmmm, no hub here.

File: provisioning/device/samples/register_x509.js [hunk 7]
```diff
@@ -5,7 +5,13 @@
 
 var fs = require('fs');
 var Transport = require('azure-iot-provisioning-device-http').Http;
+
+// Feel free to change the preceding using statement to anyone of the following if you would like to try another protocol.
```
Comment (anthonyvercolano) [inline provisioning/device/samples/register_x509.js]: >preceding using statement to anyone [](start = 27, length = 35)

preceeding ?!using!? statement to any of the following

The using doesn't belong.
Might change it to something about commenting the previous statement and uncommenting one of the subsequent statements to try another protocol.

File: provisioning/device/src/client.ts [hunk 8]
```diff
@@ -7,6 +7,9 @@ import { TpmRegistration } from './tpm_registration';
 import { RegistrationClient, X509ProvisioningTransport, TpmProvisioningTransport, X509SecurityClient, TpmSecurityClient } from './interfaces';
 import { errors } from 'azure-iot-common';
 
+/**
+ * Client object used to communicate with the Azure IoT Hub Device Provisioning Service.
```
Comment (anthonyvercolano) [inline provisioning/device/src/client.ts]: >Hub [](start = 56, length = 3)

no like hub.

**Comment by anthonyvercolano** [comment]
>   cancel(callback: (err: Error) => void): void;

Need to update these signatures.  Also, I have no idea about what to do with this cancel function.

---
Refers to: provisioning/device/src/interfaces.ts:193 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

File: provisioning/transport/amqp/readme.md [hunk 9]
```diff
@@ -1,13 +1,22 @@
...
-Communicate with Azure Device Provisioning Service over AMQP
+This package (**azure-iot-provisioning-device-amqp**) provides the AMQP and AMQP over Websocket transports that can be used to communicate with the Azure IoT Hub Device Provisioning Service. Please note that the AMQP transport supports both X509 and TPM authentication.
 
-[![npm version](https://badge.fury.io/js/azure-iot-provisioning-amqp.svg)](https://badge.fury.io/js/azure-iot-provisioning-amqp)
+[![npm version](https://badge.fury.io/js/azure-iot-provisioning-device.svg)](https://badge.fury.io/js/azure-iot-provisioning-device)
 
 ## Install
 
-`npm install -g azure-iot-provisioning-amqp@latest` to get the latest (pre-release) version.
+There are at least 3 different packages involved when writing code to take advantage of the Azure IoT Hub Device Provisioning Service:
```
Comment (anthonyvercolano) [inline provisioning/transport/amqp/readme.md]: > Hub [](start = 101, length = 4)

no hub.

File: provisioning/transport/amqp/readme.md [hunk 10]
```diff
@@ -1,13 +1,22 @@
-#azure-iot-provisioning-amqp
-Communicate with Azure Device Provisioning Service over AMQP
+This package (**azure-iot-provisioning-device-amqp**) provides the AMQP and AMQP over Websocket transports that can be used to communicate with the Azure IoT Hub Device Provisioning Service. Please note that the AMQP transport supports both X509 and TPM authentication.
```
Comment (anthonyvercolano) [inline provisioning/transport/amqp/readme.md]: >Hub [](start = 158, length = 3)

hub.

**Comment by anthonyvercolano** [comment]
>             // TODO: this only works for x509

TODO!!!

---
Refers to: provisioning/transport/amqp/src/amqp.ts:92 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

File: provisioning/transport/http/readme.md [hunk 11]
```diff
@@ -1,13 +1,22 @@
-#azure-iot-provisioning-http
-Communicate with Azure Device Provisioning Service over HTTP
+This package (**azure-iot-provisioning-device-http**) provides the HTTP transports that can be used to communicate with the Azure IoT Hub Device Provisioning Service. Please note that the HTTP transport supports both X509 and TPM authentication.
```
Comment (anthonyvercolano) [inline provisioning/transport/http/readme.md]: >Hub [](start = 134, length = 3)

Hub here and down below.

File: provisioning/transport/mqtt/readme.md [hunk 12]
```diff
@@ -1,13 +1,22 @@
-#azure-iot-provisioning-mqtt
-Communicate with Azure Device Provisioning Service over MQTT
+This package (**azure-iot-provisioning-device-mqtt**) provides the MQTT and MQTT over Websocket transports that can be used to communicate with the Azure IoT Hub Device Provisioning Service. Please note that the MQTT transport only supports X509 authentication.
```
Comment (anthonyvercolano) [inline provisioning/transport/mqtt/readme.md]: > Hub [](start = 157, length = 4)

Hub here and below.

**Comment by anthonyvercolano** [comment]
> import { translateError } from 'azure-iot-provisioning-device';

We're not consistent in our ordering of imports.   I am certainly an offender in this regard.

---
Refers to: provisioning/transport/mqtt/src/mqtt.ts:17 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Comment by anthonyvercolano** [comment]
>   activateSymmetricIdentity(callback: (err?: Error) => void): void {

They don't match the names in the interfaces.ts file.

---
Refers to: security/tpm/src/tpm.ts:23 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

File: security/x509/readme.md [hunk 13]
```diff
@@ -1,13 +1,22 @@
-#azure-iot-security-x509
-Security module for Azure IoT x509 security
+Security module used to provide X509 authentication capabilities to the Azure IoT Hub device client and Azure IoT Hub Provisioning Service device client.
```
Comment (anthonyvercolano) [inline security/x509/readme.md]: > Hub [](start = 113, length = 4)

Hub here and below.

File: common/core/src/authorization.ts (hunk 2 above)
Comment (pierreca) [reply common/core/src/authorization.ts]: looks like it's the regex on line 24

---
In reply to: [157662646](https://github.com/Azure/azure-iot-sdk-node/pull/184#discussion_r157662646) [](ancestors = 157662646)

File: provisioning/device/readme.md (hunk 4 above)
Comment (pierreca) [reply provisioning/device/readme.md]: yup

---
In reply to: [157664937](https://github.com/Azure/azure-iot-sdk-node/pull/184#discussion_r157664937) [](ancestors = 157664937)

File: provisioning/device/samples/register_x509.js (hunk 7 above)
Comment (pierreca) [reply provisioning/device/samples/register_x509.js]: supposed to be 'require'. fixed.

---
In reply to: [157665622](https://github.com/Azure/azure-iot-sdk-node/pull/184#discussion_r157665622) [](ancestors = 157665622)

**Comment by pierreca** [comment]
>   cancel(callback: (err: Error) => void): void;

That's fine. everything TPM is private for now, we can change it later.

---
In reply to: [352641806](https://github.com/Azure/azure-iot-sdk-node/pull/184#issuecomment-352641806) [](ancestors = 352641806)

---
Refers to: provisioning/device/src/interfaces.ts:193 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Comment by pierreca** [comment]
>             // TODO: this only works for x509

we'll get there when we implement TPM. I'm removing the TODO comment for clean up but in theory it should work fine since TPM is only used for initial authentication but not during the registration flow itself.

---
In reply to: [352642730](https://github.com/Azure/azure-iot-sdk-node/pull/184#issuecomment-352642730) [](ancestors = 352642730)

---
Refers to: provisioning/transport/amqp/src/amqp.ts:92 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Comment by pierreca** [comment]
> import { translateError } from 'azure-iot-provisioning-device';

we can do a pass on that later. won't fix based on the fact that it's unrelated to the change - good feedback though. 

---
In reply to: [352643429](https://github.com/Azure/azure-iot-sdk-node/pull/184#issuecomment-352643429) [](ancestors = 352643429)

---
Refers to: provisioning/transport/mqtt/src/mqtt.ts:17 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Comment by pierreca** [comment]
>   private _credentials: TransportConfig;

name seemed explicit enough in my opinon but feel free to discuss/suggest something.

---
In reply to: [352638700](https://github.com/Azure/azure-iot-sdk-node/pull/184#issuecomment-352638700) [](ancestors = 352638700)

---
Refers to: device/core/src/sak_authentication_provider.ts:23 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Comment by pierreca** [comment]
>   activateSymmetricIdentity(callback: (err?: Error) => void): void {

that's ok we're not shipping this package yet so it can change.

---
In reply to: [352643679](https://github.com/Azure/azure-iot-sdk-node/pull/184#issuecomment-352643679) [](ancestors = 352643679)

---
Refers to: security/tpm/src/tpm.ts:23 in ca31583. [](commit_id = ca31583e7dbdc983d4bc11b5847d3be58659d394, deletion_comment = False)

**Pull request merged**